src/Microsoft.Performance.SDK.Runtime.NetCoreApp/Plugins/PluginsLoader.cs

```diff
@@ -235,26 +241,59 @@ public async Task<(bool, IDictionary<string, ErrorInfo>)> TryLoadPluginAsync(IEn
         /// <exception cref="ObjectDisposedException">
         ///     This instance is disposed.
         /// </exception>
+        /// <exception cref="ArgumentNullException">
+        ///     Either <paramref name="directories"/> or one of its elements is null.
+        /// </exception>
         public bool TryLoadPlugins(IEnumerable<string> directories, out IDictionary<string, ErrorInfo> failed)
         {
+            Guard.NotNull(directories, nameof(directories));
+            foreach (string dir in directories)
+            {
+                Guard.NotNull(dir, nameof(directories));
+            }
+            
+            failed = new Dictionary<string, ErrorInfo>();
+
+            if (!directories.Any())
+            {
+                return true;
+            }
+
             this.ThrowIfDisposed();
             lock (this.mutex)
             {
-                failed = new Dictionary<string, ErrorInfo>();
                 var oldPlugins = new HashSet<ProcessingSourceReference>(this.extensionRoot.ProcessingSources);
                 foreach (var dir in directories)
                 {
+                    ErrorInfo emptyErrorInfo = new ErrorInfo(ErrorCodes.InvalidArgument, ErrorCodes.InvalidArgument.Description)
+                    {
+                        Target = dir,
+                    };
+
+                    Guard.NotNull(dir, nameof(directories));
+                    
+                    if (string.IsNullOrWhiteSpace(dir))
+                    {
+                        failed.Add(dir, emptyErrorInfo);
+                        continue;
+                    }
+                    
                     if (!this.extensionDiscovery.ProcessAssemblies(dir, out var error))
                     {
                         failed.Add(dir, error);
                         continue;
                     }
                 }
-
-                // TODO: this does redundant work re-finalizing processing sources
-                // loaded by previous calls to this method. The extension repository
-                // should get refactored to avoid this redundant work.
-                this.extensionRoot.FinalizeDataExtensions();
+                
+                // If nothing was able to be loaded above, skip finalizing 
+                if (this.extensionRoot.IsLoaded)
+                {
+                    // TODO: this does redundant work re-finalizing processing sources
```

Comment:
Do we have an issue open for this? Can we add one, otherwise?

Reply:
Sure